### ISSUE TITLE
TypeLoadException while mocking mixed code interface that contains a method with a ref parameter of type unsigned long.

### ISSUE THREAD

**Issue by BillArmstrong** [moq/moq4#670]
This issue is identical to issue #244 that was closed. The only difference is that the parameter type is different.

The following interface will throw the exception:

`	public interface class ITestInterface
	{
		virtual void GetValue(unsigned long %value);
	};
`
If I change it from "unsigned long" to "unsigned int", it will work with no exceptions:

`	public interface class ITestInterface
	{
		virtual void GetValue(unsigned int %value);
	};
`

In C++ the data types "unsigned long" and "unsigned int" are identical. They both correspond to UInt32. I'm not sure why one works and the other does not.

**Comment by stakx** [comment]
@BillArmstrong: I have no C++/CLI compiler available right now. Could you please provide...

* a compiled test assembly containing two such `public` test interfaces (one with `unsigned int`, the other with `unsigned long`)?

* the Moq version number you're using

* an indication of which .NET platform & verdion you're running your code on?

Thanks!

**Comment by stakx** [comment]
In the meantime I'll make a first blind guess that [this](https://github.com/dotnet/coreclr/pull/17881#issuecomment-387114966) would be a possible cause of this problem.

**Comment by stakx** [comment]
**TL;DR:** This bug is due to a limitation of .NET Reflection, upon which Moq 4 is built. There's nothing we can do about it, so I'll close this. If you're interested in some details, read on.

Unfortunately, my above guess was correct.

I compiled a C++/CLI class library with the following two interfaces:

```c++
public interface class IWithUnsignedInt { virtual void GetValue(unsigned int %value); };
public interface class IWithUnsignedLong { virtual void GetValue(unsigned long %value); };
```

Then I inspected the generated IL metadata using ILDASM. In IL, these methods have the following signatures:

```
// IWithUnsignedInt.GetValue (raw signature bytes: 20 01 01 10 09)
public hidebysig newslot abstract virtual instance void GetValue(
    uint32& 'value') cil managed

// IWithUnsignedLong.GetValue (raw signature bytes: 20 01 01 10 20 41 09)
public hidebysig newslot abstract virtual instance void GetValue(
    uint32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong)& 'value') cil managed
```

Notice how the signatures differ by an optional custom modifier (the `modopt` bit / the bytes `20 41` in the raw method signature).

Now that we know for a fact that the parameter of `IWithUnsignedLong.GetValue` carries a modopt, let's see if we can read it using Reflection:

```csharp
var parameter = typeof(IWithUnsignedLong).GetMethod("GetValue").GetParameters()[0];
var modopts = parameter.GetOptionalCustomModifiers();
Assert.NotEmpty(modopts);
```

This assertion fails, meaning that .NET Reflection does not report any optional custom modifiers although one is present. Because of this, the mocked method produced by Moq (or more precisely, by DynamicProxy) will have a different signature than the original, but at the same time claim to be an override of the latter&mdash;which, from the runtime's point of view, is not possible due to the signature mismatches. That's why you get a `MissingMethodException` (or similar).

[@AtsushiKan's post](https://github.com/dotnet/coreclr/pull/17881#issuecomment-387114966) (which I referred to in my previous post) implied why .NET Reflection doesn't do the right thing here: The modopt is not placed on the full, "outer" by-ref type `uint&`, but only on the "nested" element type `uint`, so Reflection doesn't see it:

> [T]his will propagate custom modifiers attached to the parameter type itself but not any types are used to construct the type. E.g. a parameter of type `int[]` can have a custom modifier attached to the `int[]` or the `int` - you'll only propagate the former.
>
> This is a long standing defect in how Reflection thinks about custom modifiers and it's not likely you could address this here without majorly impacting the cost of the fix. Just noting it here so we know the limitation exists [...].

Sorry for the bad news. Closing this as unfixable.
